great_expectations/zep/metadatasource.py

```diff
@@ -27,30 +32,31 @@ def __new__(
 
         Also binds asset adding methods according to the declared `asset_types`.
         """
-        LOGGER.info(f"1a. {meta_cls.__name__}.__new__() for `{cls_name}`")
-
-        # TODO: extract asset type details to build factory method signature etc. (pull args from __init__)
-
-        asset_types: List[Type[DataAsset]] = cls_dict.get("asset_types")
-        LOGGER.info(f"1b. Extracting Asset details - {asset_types}")
+        LOGGER.debug(f"1a. {meta_cls.__name__}.__new__() for `{cls_name}`")
 
         cls = super().__new__(meta_cls, cls_name, bases, cls_dict)
-        LOGGER.debug(f"  {cls_name} __dict__ ->\n{pf(cls.__dict__, depth=3)}")
 
         if cls_name == "Datasource":
             # NOTE: the above check is brittle and must be kept in-line with the Datasource.__name__
-            LOGGER.info("1c. Skip factory registration of base `Datasource`")
+            LOGGER.debug("1c. Skip factory registration of base `Datasource`")
             return cls
 
-        def _datasource_factory(*args, **kwargs) -> Datasource:
+        LOGGER.debug(f"  {cls_name} __dict__ ->\n{pf(cls.__dict__, depth=3)}")
+
+        meta_cls.__cls_set.add(cls)
+        LOGGER.info(f"Datasources: {len(meta_cls.__cls_set)}")
+
+        def _datasource_factory(name: str, **kwargs) -> Datasource:
             # TODO: update signature to match Datasource __init__ (ex update __signature__)
-            LOGGER.info(f"5. Adding `{args[0] if args else ''}` {cls_name}")
-            return cls(*args, **kwargs)
+            LOGGER.info(f"5. Adding '{name}' {cls_name}")
+            return cls(name=name, **kwargs)
 
         # TODO: generate schemas from `cls` if needed
 
-        _SourceFactories.register_factory(
-            cls, _datasource_factory, asset_types=asset_types
-        )
+        if cls.__module__ == "__main__":
+            LOGGER.warning(
```

Comment:
Sometimes it causes a collision and is a problem, sometimes it isn't. 
I think it's related to what else is being imported and run but I didn't spend the time to isolate it since it's a contrived condition.